pkg/gql/query/blocks.go

```diff
@@ -0,0 +1,208 @@
+
+	core "github.com/dusk-network/dusk-blockchain/pkg/core/transactions"
+)
+
+// File purpose is to define all arguments and resolvers relevant to "blocks" query only
+
+type blocks struct {
+	db database.DB
+}
+
+func (b blocks) getQuery() *graphql.Field {
+	return &graphql.Field{
+		Type: graphql.NewList(Block),
+		Args: graphql.FieldConfigArgument{
+			"hash": &graphql.ArgumentConfig{
+				Type: graphql.String,
+			},
+			"hashes": &graphql.ArgumentConfig{
+				Type: graphql.NewList(graphql.String),
+			},
+			"height": &graphql.ArgumentConfig{
+				Type: graphql.Int,
+			},
+			"range": &graphql.ArgumentConfig{
+				Type: graphql.NewList(graphql.Int),
+			},
+			"last": &graphql.ArgumentConfig{
+				Type: graphql.Int,
+			},
+		},
+		Resolve: b.resolve,
+	}
+}
+
+func (b blocks) resolve(p graphql.ResolveParams) (interface{}, error) {
+
+	// resolve argument hash (single block)
+	hash, ok := p.Args["hash"].(interface{})
+	if ok {
+		hashes := make([]interface{}, 0)
+		hashes = append(hashes, hash)
+		return b.fetchBlocksByHashes(hashes)
+	}
+
+	// resolve argument hashes (multiple blocks)
+	hashes, ok := p.Args["hashes"].([]interface{})
+	if ok {
+		return b.fetchBlocksByHashes(hashes)
+	}
+
+	// resolve argument height (single block)
+	height, ok := p.Args["height"].(int)
+	if ok {
+		return b.fetchBlocksByHeights(int64(height), int64(height))
+	}
+
+	// resolve argument range (range of blocks)
+	heightRange, ok := p.Args["range"].([]interface{})
+	if ok && len(heightRange) == 2 {
+		from, ok := heightRange[0].(int)
+		if !ok {
+			return nil, errors.New("range `from` value not int64")
+		}
+
+		to, ok := heightRange[1].(int)
+		if !ok {
+			return nil, errors.New("range `to` value not int64")
+		}
+
+		return b.fetchBlocksByHeights(int64(from), int64(to))
+	}
+
+	offset, ok := p.Args["last"].(int)
+	if ok {
+		if offset <= 0 {
+			return nil, errors.New("invalid offset")
+		}
+		return b.fetchBlocksByHeights(int64(offset)*-1, -1)
+	}
+
+	return nil, nil
+}
+
+func resolveTxs(p graphql.ResolveParams) (interface{}, error) {
+
+	var txs []core.Standard
+	b, ok := p.Source.(*block.Block)
+	if ok {
+		// TODO: use the one from the context
+		_, db := heavy.CreateDBConnection()
+		err := db.View(func(t database.Transaction) error {
+			fetched, err := t.FetchBlockTxs(b.Header.Hash)
+			if err != nil {
+				return err
+			}
+
+			for _, tx := range fetched {
+				sTx := tx.StandardTX()
+				sTx.TxID, _ = tx.CalculateHash()
+				txs = append(txs, sTx)
+			}
+			return nil
+		})
+
+		return txs, err
+	}
+	return nil, errors.New("invalid source block")
+}
+
+// Fetch block headers by a list of hashes
+func (b blocks) fetchBlocksByHashes(hashes []interface{}) ([]*block.Block, error) {
+
+	blocks := make([]*block.Block, 0)
+	err := b.db.View(func(t database.Transaction) error {
+		for _, v := range hashes {
+			encodedHash, ok := v.(string)
+			if !ok {
+				continue
+			}
+
+			hash, err := base64.StdEncoding.DecodeString(encodedHash)
```

Comment:
Maybe it would be better to take in hex strings, instead of base64, as hex strings are the common encoding for most tx/block related information.

Reply:
Yes. Maybe it would make sense to use a custom type for `hash` with a predefined `String()` method that Hex encodes the `[]byte`?

Reply:
> Maybe it would be better to take in hex strings, instead of base64, as hex strings are the common encoding for most tx/block related information.

Indeed! I chose Base64 as it gives a shorter representation than Hex. But as we already have the wallet working with Hex encoding, better to be aligned with that one for now.

Reply:
> Yes. Maybe it would make sense to use a custom type for hash with a predefined String() method that Hex encodes the []byte?

Better to handle this with another task where core packages can be updated accordingly too.

Reply:
Agreed, feel free to open an issue.